src/models/highlightedtext.cpp

```diff
@@ -0,0 +1,281 @@
+{
+    QVector<QTextLayout::FormatRange> formats;
+
+    int offset = 0;
+    QTextLayout::FormatRange format;
+
+    const auto escapeChar = QLatin1Char('\u001B');
+
+    constexpr int setColorSequenceLength = 5;
+    constexpr int resetColorSequenceLength = 4;
+    constexpr int colorCodeLength = 2;
+
+    for (const auto& line : text) {
+        auto lastToken = line.cbegin();
+        int lineOffset = 0;
+        for (auto escapeIt = std::find(line.cbegin(), line.cend(), escapeChar); escapeIt != line.cend();
+             escapeIt = std::find(escapeIt, line.cend(), escapeChar)) {
+
+            lineOffset += std::distance(lastToken, escapeIt);
+            Q_ASSERT(*(escapeIt + 1) == QLatin1Char('['));
+
+            // escapeIt + 2 points to the first color code character
+            auto color = QStringView {escapeIt + 2, colorCodeLength};
+            bool ok = false;
+            const uint8_t colorCode = color.toUInt(&ok);
+            if (ok) {
+                // only support the 8 default colors
+                Q_ASSERT(colorCode >= 30 && colorCode <= 37);
+
+                format.start = offset + lineOffset;
+                const auto colorRole = static_cast<KColorScheme::ForegroundRole>(colorCode - 30);
+                format.format.setForeground(m_colorScheme.foreground(colorRole));
+
+                std::advance(escapeIt, setColorSequenceLength);
+            } else {
+                // make sure we have a reset sequence
+                Q_ASSERT(color == QStringLiteral("0m"));
+                format.length = offset + lineOffset - format.start;
+                if (format.length) {
+                    formats.push_back(format);
+                }
+
+                std::advance(escapeIt, resetColorSequenceLength);
+            }
+            lastToken = escapeIt;
+        }
+        offset += lineOffset + std::distance(lastToken, line.cend());
+    }
+
+    return formats;
+}
+
+void AnsiHighlightingImplementation::themeChanged()
+{
+    m_colorScheme = KColorScheme(QPalette::Normal, KColorScheme::Complementary);
+}
+
+void AnsiHighlightingImplementation::setHighlightingDefinition(const KSyntaxHighlighting::Definition& /*definition*/) {
+}
+
+QString AnsiHighlightingImplementation::definitionName() const
+{
+    return {};
+}
+
+HighlightedText::HighlightedText(KSyntaxHighlighting::Repository* repository, QObject* parent)
+    : QObject(parent)
+#if KFSyntaxHighlighting_FOUND
+    , m_repository(repository)
+#endif
+    , m_layout(std::make_unique<QTextLayout>())
+{
+    m_layout->setCacheEnabled(true);
+    m_layout->setFont(QFontDatabase::systemFont(QFontDatabase::FixedFont));
+}
+
+HighlightedText::~HighlightedText() = default;
+
+void HighlightedText::setText(const QStringList& text)
+{
+    const bool usesAnsi = std::any_of(text.cbegin(), text.cend(),
+                                      [](const QString& line) { return line.contains(QLatin1Char('\u001B')); });
```

Comment:
again this magic char, what does it mean? if it has to be in sync with the other place, introduce it as a constexpr global with a sane name

Reply:
well its know as escape character
if you look at the ascii table you can see it there
```
       032   26    1A    SUB (substitute)            132   90    5A    Z
       033   27    1B    ESC (escape)                133   91    5B    [
       034   28    1C    FS  (file separator)        134   92    5C    \  '\\'
```

Reply:
you cannot expect people have the ASCII table in their mind when reading the code. the point is: your code is unreadable as-is. make it readable by introducing a named variable:

```
constexpr auto escapeCode = QLatin1Char('\u001B');
```

Reply:
you even introduced this now (in the wrong patch) - use it here too

Reply:
this isn't done